services/src/main/java/org/keycloak/authentication/authenticators/browser/IdentityProviderAuthenticator.java

```diff
@@ -58,6 +59,16 @@ public void authenticate(AuthenticationFlowContext context) {
                 redirect(context, providerId);
             }
         } else if (context.getAuthenticatorConfig() != null && context.getAuthenticatorConfig().getConfig().containsKey(IdentityProviderAuthenticatorFactory.DEFAULT_PROVIDER)) {
+            if (context.getForwardedErrorMessage() != null)
+            {
+                LOG.infof("Should redirect to remote IdP but forwardedError has value '%s', redirecting to error page", context.getForwardedErrorMessage());
+
+                LoginFormsProvider forms = context.getSession().getProvider(LoginFormsProvider.class).setAuthenticationSession(context.getAuthenticationSession());
+                forms.addError(context.getForwardedErrorMessage());
+                context.forceChallenge(forms.createErrorPage(Response.Status.BAD_REQUEST));
```

Comment:
Will it be better to use "context.challenge" here? I am not 100% sure, but ideally this would mean that:
- In case that IdentityProviderAuthenticator is REQUIRED, the challenge will be forced to be displayed
- In case that IdentityProviderAuthenticator is ALTERNATIVE, the challenge won't be displayed, but it will go to the next authenticator (usually username&password screen) and the error will be displayed on this screen. Which will give user a chance to fallback to the authentication with username&password . WDYT?

Reply:
I agree your proposal is better, however after trying it seems that the `context.challenge` call the control is immediately passed to the challenge (error) page and it never gets a chance to reach the forms authenticator.

I have also tried replacing this code with a simple context.attempted(), when ALTERNATIVE this actually redirects to the forms authenticator and shows the error message, but when REQUIRED it goes directly to the "Invalid username or password" screen and does not show the forwarded error. Maybe the issue is that forwarded messages are lost when the flow ends without any successful execution?

Reply:
Hm... I am not 100% sure. It is possible that it is a bug as "context.challenge" should ideally work in same way as "context.attempted" . The only difference is, that "context.challenge" provides a way for the authenticator to provision the screen when the "context.attempted" is mostly targetted for non-interactive authenticators when showing any specific screen is usually not desired. But maybe I am missing something...

If you want to investigate the authenticationFlow (and possibly fix at the authenticationFlow level), it will be ideal. But if not, it is ok for now to keep things as you did in PR. Just maybe:
- add TODO in the code and possibly refer to this PR (or summary in the TODO if possible)
- Additional very minor thing: Could you please use:
```
if (something) {
```
instead of:
```
if (something)
{
```

And finally the test :-)
Thanks!

Reply:
Yep, the tests are always the hardest part. I still have no idea how to setup the realms and config, usually I just go look for some test similar to the one I need and start to tweak, but it takes ages. Will try anyway! :)

Regarding the .challenge issue, I think it's because the flow executes the challenge with the same priority of the authenticator, and the IdP auth has a greater priority than the forms auth. I'll check if I can change the challenge priority. Any suggestions welcome, and thanks for reviewing this!

Reply:
Updated with a slightly different approach. I'll mark the thread as resolved and wait for your kind review on the other one.